autogpts/autogpt/autogpt/commands/web_search.py

```diff
@@ -1,169 +1,191 @@
-"""Commands to search the web with"""
-
-from __future__ import annotations
-
 import json
+import logging
 import time
+from typing import Iterator
 
 from duckduckgo_search import DDGS
 
-from autogpt.agents.agent import Agent
-from autogpt.agents.utils.exceptions import ConfigurationError
+from autogpt.agents.protocols import CommandProvider, DirectiveProvider
 from autogpt.command_decorator import command
+from autogpt.config.config import Config
 from autogpt.core.utils.json_schema import JSONSchema
+from autogpt.models.command import Command
+from autogpt.utils.exceptions import ConfigurationError
 
 COMMAND_CATEGORY = "web_search"
 COMMAND_CATEGORY_TITLE = "Web Search"
 
-
 DUCKDUCKGO_MAX_ATTEMPTS = 3
 
+logger = logging.getLogger(__name__)
 
-@command(
-    "web_search",
-    "Searches the web",
-    {
-        "query": JSONSchema(
-            type=JSONSchema.Type.STRING,
-            description="The search query",
-            required=True,
-        )
-    },
-    aliases=["search"],
-)
-def web_search(query: str, agent: Agent, num_results: int = 8) -> str:
-    """Return the results of a Google search
-
-    Args:
-        query (str): The search query.
-        num_results (int): The number of results to return.
 
-    Returns:
-        str: The results of the search.
-    """
-    search_results = []
-    attempts = 0
+class WebSearchComponent(DirectiveProvider, CommandProvider):
+    """Provides commands to search the web."""
 
-    while attempts < DUCKDUCKGO_MAX_ATTEMPTS:
-        if not query:
-            return json.dumps(search_results)
+    def __init__(self, config: Config):
+        self.legacy_config = config
 
-        search_results = DDGS().text(query, max_results=num_results)
+    def get_resources(self) -> Iterator[str]:
+        yield "Internet access for searches and information gathering."
 
-        if search_results:
-            break
+    def get_commands(self) -> Iterator[Command]:
+        yield self.web_search
 
-        time.sleep(1)
-        attempts += 1
+        if (
+            self.legacy_config.google_api_key
+            and self.legacy_config.google_custom_search_engine_id
+        ):
+            yield self.google
+        else:
+            logger.info("Configure google_api_key and custom_search_engine_id.")
```

Comment:
This (+ a few similar messages) is now logged every single cycle instead of at startup. Quite annoying :/

Reply:
I moved them to `__init__`